hilti/toolchain/src/compiler/codegen/codegen.cc

```diff
@@ -149,7 +149,8 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
     void operator()(const declaration::Type& n, position_t p) {
         assert(n.typeID());
         cg->compile(n.type(), codegen::TypeUsage::Storage);
-        cg->addTypeInfoDefinition(n.type());
+        if ( include_implementation )
+            cg->addTypeInfoDefinition(n.type());
```

Comment:
I revisited the example we had also looked at offline,

``` $ spicyc -C tests/spicy/types/real/coercion.spicy -j```

Before #984 we emitted only `dbg.Foo.cc` (w/o any type infos), and `dbg.__linker__.cc` (w/o type infos). After #984 we emitted an additional `dbg.hilti.cc` which included type infos in addition to now also emitted type infos in `dbg.Foo.cc`. The hypothesis was that the additional type infos slowed down the compilation.

Now looking at this patch I see that the type infos in `dbg.Foo.cc` are gone and we are pretty much back to the state before #984 which is great. However, I still see `dbg.hilti.cc` being emitted, with type infos which seem unused. This seems to be due to how we attach type infos to units here and below, so we do not check from the top-level user code which types are used, but instead bubble them up from whatever code we see. I suspect that us emitted that additional `.cc` file for hilti still contributes significantly to the overall time it takes to compile, not only due to the extra type infos, but more so since it requires us to start another compilation job.

Do you think we could collect type infos at a higher level were we can actually check whether it would be possible to reference them in generated C++ code?

Reply:
Typeinfos aren't represented in the AST at all currently, so I suspect that would be a major undertaking. Let me see if I can make it stop generating that additional compilation unit, like it used to be.